internal/server/stat_point.go

```diff
@@ -155,80 +155,121 @@ func (s *Server) GetStatSetWithinPlace(
+			Metadata: make(map[string]*pb.StatMetadata),
 		}
-		for _, statVar := range statVars {
-			result.Data[statVar] = &pb.PlacePointStat{
-				Stat:     make(map[string]*pb.PointStat),
-				Metadata: make(map[string]*pb.StatMetadata),
-			}
+	}
+
+	// Read from cache directly
+	rowList, keyTokens := buildStatSetWithinPlaceKey(parentPlace, childType, dateKey, statVars)
+	cacheData, err := readStatCollection(ctx, s.store, rowList, keyTokens)
+	if err != nil {
+		return nil, err
+	}
+
+	gotResult := false
+	for statVar, data := range cacheData {
+		if data == nil {
+			continue
 		}
-		gotResult := false
-		for statVar, data := range cacheData {
-			if data != nil {
-				gotResult = true
-				cohorts := data.SourceCohorts
-				// Sort cohort first, so the preferred source is populated first. Then only
-				// update when there is a later data.
-				sort.Sort(SeriesByRank(cohorts))
-				for _, cohort := range cohorts {
-					for place, val := range cohort.Val {
-						pointStat, ok := result.Data[statVar].Stat[place]
-						if !ok || pointStat.Date < cohort.PlaceToLatestDate[place] {
-							metaData := &pb.StatMetadata{
-								MeasurementMethod: cohort.MeasurementMethod,
-								ObservationPeriod: cohort.ObservationPeriod,
-								ProvenanceUrl:     cohort.ProvenanceUrl,
-								ScalingFactor:     cohort.ScalingFactor,
-								ImportName:        cohort.ImportName,
-								Unit:              cohort.Unit,
-							}
-							result.Data[statVar].Stat[place] = &pb.PointStat{
-								Date:     cohort.PlaceToLatestDate[place],
-								Value:    val,
-								Metadata: metaData,
-							}
-							result.Data[statVar].Metadata[cohort.ImportName] = metaData
-						}
+		gotResult = true
+		cohorts := data.SourceCohorts
+		// Sort cohort first, so the preferred source is populated first.
+		sort.Sort(SeriesByRank(cohorts))
+		// update when there is a later data.
+		for _, cohort := range cohorts {
+			for place, val := range cohort.Val {
+				pointStat, ok := result.Data[statVar].Stat[place]
+				// This works when date is set. The result will be populated in first
+				// for loop only.
+				if !ok || pointStat.Date < cohort.PlaceToLatestDate[place] {
+					metaData := &pb.StatMetadata{
+						MeasurementMethod: cohort.MeasurementMethod,
+						ObservationPeriod: cohort.ObservationPeriod,
+						ProvenanceUrl:     cohort.ProvenanceUrl,
+						ScalingFactor:     cohort.ScalingFactor,
+						ImportName:        cohort.ImportName,
+						Unit:              cohort.Unit,
+					}
+					usedDate := date
+					if usedDate == "" {
+						usedDate = cohort.PlaceToLatestDate[place]
+					}
+					result.Data[statVar].Stat[place] = &pb.PointStat{
+						Date:     usedDate,
+						Value:    val,
+						Metadata: &pb.StatMetadata{ImportName: cohort.ImportName},
 					}
+					result.Data[statVar].Metadata[cohort.ImportName] = metaData
 				}
 			}
 		}
-		if gotResult {
-			return result, nil
+	}
+	// Check if need to read from memory database.
```

Comment:
nit: this was already a long function doing two things, and we're adding a third, might be good to modularize it into functions and comment like 1. read cache x, 2. read memdb, 3. read cache y, ...

Reply:
Okay, I will do this in a follow up PR.